upstairs/src/downstairs.rs

```diff
@@ -2474,6 +2538,13 @@ impl Downstairs {
                 "downstairs failed, too many outstanding bytes {byte_count}"
             );
             Some(ClientStopReason::TooManyOutstandingBytes)
+        } else if !self.can_replay {
+            // XXX can this actually happen?
+            warn!(
+                self.log,
+                "downstairs became ineligible for replay while offline"
```

Comment:
I don't think it's possible, but the state machines are complex enough that I'm not 100% certain; faulting the downstairs is a correct behavior, so I'd rather do that versus panicking.  I did try running unit tests with a panic here, and it didn't fire, but that could just mean that it's an unlikely corner case.

> It also seems like the statement downstairs became ineligible for replay while offline is not the unusual part, is it? If a downstairs is Offline and we continue to get IO, then it will become ineligible for replay and that's kind of the expected behavior, though maybe not in this method?

The `can_replay` flag is set to `false` when we remove _completed_ jobs that would be needed for a replay ("completed" = done / skipped / error on all three downstairs).  That shouldn't happen with one downstairs Offline, because jobs are left `InProgress` for the offline downstairs (for exactly this reason).

(We can also become ineligible for replay when too many `InProgress` jobs pile up, which is checked right here)